test/test_svd.cc

```diff
@@ -208,7 +216,61 @@ void test_svd_work( Params& params, bool run )
         }
     }
 
-    if (check || ref) {
+    if (check && (wantu || wantvt)) {
+        // Residual matrix.
+        slate::Matrix<scalar_t> R( min_mn, min_mn, nb, p, q, MPI_COMM_WORLD );
+        R.insertLocalTiles();
+
+        if (wantu) {
+            //==================================================
+            // Test results by checking orthogonality of U
+            //
+            //      || I - U^H U ||_1
+            //     ------------------- < tol * epsilon
+            //              N
+            //==================================================
+            slate::set( zero, one, R ); // identity
+            auto UH = conj_transpose( U );
+            slate::gemm( -one, UH, U, one, R );
+            params.ortho_U() = slate::norm( slate::Norm::One, R ) / n;
+            params.okay() = params.okay() && (params.ortho_U() <= tol);
+        }
+
+        if (wantvt) {
+            //==================================================
+            // Test results by checking orthogonality of VT
+            //
+            //      || I - V V^H ||_1
+            //     ------------------- < tol * epsilon
+            //              N
+            //==================================================
+            slate::set( zero, one, R ); // identity
+            auto V = conj_transpose( VT );
+            slate::gemm( -one, VT, V, one, R );
+            params.ortho_V() = slate::norm( slate::Norm::One, R ) / n;
+            params.okay() = params.okay() && (params.ortho_V() <= tol);
+        }
+
+        if (wantu && wantvt) {
+            //==================================================
+            // Test results by checking backwards error
+            //
+            //      || Acpy - U Sigma VT ||_1
+            //     --------------------------- < tol * epsilon
+            //            || A ||_1 * N
+            //
+            //==================================================
+            // todo:
+            slate::scale_row_col( slate::Equed::Col, Sigma, Sigma, U );
+
+            real_t Anorm = slate::norm( slate::Norm::One, Acpy );
+            slate::gemm( -one, U, VT, one, Acpy );
+            params.error2() = slate::norm( slate::Norm::One, Acpy ) / (Anorm * n);
+            params.okay() = params.okay() && (params.error2() <= tol);
+        }
+    }
+
+    if (ref) {
```

Comment:
Should this still be `ref || check`? (Or should `check` imply `ref`?)  Otherwise, when `ref`, `jobu`, and `jobvt` are all false, no checking is done, regardless of the value of `check`.  (Maybe we can check if the matrix generator provided values, if we don't want `check` to imply `ref`.)

At a minimum, I think a warning print is needed when no checks are actually applied.

Reply:
The issue is if `--jobu s --jobvt s --check y`, then it does a backwards error check A – U Sigma V^H, which is fairly fast, but then with `if (check || ref)` it would also run ScaLAPACK, which is rather slow. I wanted to check results but not run ScaLAPACK. Unfortunately that means the `--jobu n --jobvt n --check y --ref n` case won't actually run a check; the user would have to use `--ref y`.

Printing a warning seems like a good solution. The output does say "no check", but that could be confusing to a user who set `--check y`.

Reply:
Added message. See sample output in comment below.